xc/device.ml

```diff
@@ -1989,43 +1980,42 @@ module Backend = struct
 
     (** Implementation of the Vbd functions that use the dispatcher for the qemu-upstream-compat backend *)
     module Vbd = struct
+
       let qemu_media_change ~xs device _type params =
         Vbd_Common.qemu_media_change ~xs device _type params;
-        let open Qmp in
-        let cd = "ide1-cd1" in
+        let cd    = "ide1-cd1" in
+        let domid = device.frontend.domid in
         if params = "" then
-          let qmp_cmd = Command(None, Eject (cd, Some true)) in
-          qmp_write device.frontend.domid qmp_cmd
+          qmp_send_cmd domid Qmp.(Eject(cd, Some true)) |> ignore
         else
           try
-            let c = Qmp_protocol.connect (qmp_libxl_path device.frontend.domid) in
+            let as_msg cmd = Qmp.(Success(Some __LOC__, cmd)) in
+            let fd_cd      = Unix.openfile params [ Unix.O_RDONLY ] 0o640 in
             finally
               (fun () ->
-                 let fd_of_c = Qmp_protocol.to_fd c in
-                 let fd_of_cd = Unix.openfile params [ Unix.O_RDONLY ] 0o640 in
-                 finally
-                   (fun () -> ignore(Fd_send_recv.send_fd fd_of_c " " 0 1 [] fd_of_cd))
-                   (fun () -> Unix.close fd_of_cd);
-
-                 let qmp_cmd = Command (None, Add_fd None) in
-                 Qmp_protocol.negotiate c;
-                 Qmp_protocol.write c qmp_cmd;
-                 let fd_info = match Qmp_protocol.read c with
-                   | Success (None, Fd_info x) -> x
-                   | _ -> raise (Internal_error (Printf.sprintf "Get unexpected result after sending Qmp message: %s" (string_of_message qmp_cmd)))
-                 in
+                 let cmd     = Qmp.(Add_fd None) in
+                 let fd_info = match qmp_send_cmd ~send_fd:fd_cd domid cmd with
+                   | Qmp.Fd_info x -> x
+                   | other ->
+                     raise (Internal_error
```

Comment:
I was trying to se if this could be shadowed, as it happened in the past. It does not look like the case, and I think the `Internal_error` used in here and in the rest of the module is coming from `task_server` in `xcp-idl`, at least I could not find any other definition. But I may be wrong.